### PR TITLE
Nogin Denis. Lab 3. Task 2. 

### PR DESCRIPTION
**Pull request by Gr0miLa**
Count number of machine instructions performed during execution of a function.

### CODE REVIEW

File: llvm/lib/Target/X86/CMakeLists.txt [hunk 1]
```diff
@@ -87,6 +87,7 @@ set(sources
   X86VZeroUpper.cpp
   X86WinEHState.cpp
   X86InsertWait.cpp
+  X86NoginCountInstsPass.cpp
```
Comment (m-ly4) [inline llvm/lib/Target/X86/CMakeLists.txt]: Please make this pass as dynamic library plugin.

File: llvm/lib/Target/X86/X86NoginCountInstsPass.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    const TargetInstrInfo *InstrInfo = MachFunc.getSubtarget().getInstrInfo();
+    size_t CountInstsReg =
+        MachFunc.getRegInfo().createVirtualRegister(&X86::GR64RegClass);
+    size_t Count = 0;
+
+    for (auto &BasicBlock : MachFunc) {
+      Count += countInstrBasBlock(BasicBlock);
+    }
+
+    BuildMI(MachFunc.back(), MachFunc.back().getFirstTerminator(),
```
Comment (m-ly4) [inline llvm/lib/Target/X86/X86NoginCountInstsPass.cpp]: Last block in machine functions blocks list is not necessarily the last one in real function containing return instruction.
Moreover, there may be several basic blocks with different instructions terminating function call.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/X86/X86NoginCountInstsPass.cpp (hunk 2 above)
Comment (m-ly4) [reply llvm/lib/Target/X86/X86NoginCountInstsPass.cpp]: Still there can be more than one block with termination instruction (and it is not only return) in a function.

File: llvm/lib/Target/X86/lab3/nogin_denis/X86NoginCountInstsPass.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    if (!ReturnBlock) {
+      ReturnBlock = &MachFunc.back();
+      DebugLocation = ReturnBlock->begin()->getDebugLoc();
+    }
+
+    BuildMI(*ReturnBlock, ReturnBlock->getFirstTerminator(), DebugLocation,
+            InstrInfo->get(X86::MOV64mr))
+        .addImm(Count)
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/nogin_denis/X86NoginCountInstsPass.cpp]: This is number of instructions in assembly file but not the number of really executed instructions during program run.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/X86/lab3/nogin_denis/X86NoginCountInstsPass.cpp (hunk 3 above)
Comment (Gr0miLa) [reply llvm/lib/Target/X86/lab3/nogin_denis/X86NoginCountInstsPass.cpp]: I didn't manage to do it before 12 o'clock, I hope everything is fine

**Review by m-ly4 — APPROVED** [review]

File: llvm/lib/Target/X86/lab3/nogin_denis/X86NoginCountInstsPass.cpp (hunk 3 above)
Comment (m-ly4) [reply llvm/lib/Target/X86/lab3/nogin_denis/X86NoginCountInstsPass.cpp]: Ok, lets think that you pushed last commit from London and in your timezone was 26th April yet.
But, tsss, don't tell anyone about that hack. :)

**Pull request merged**